tangelo/toolboxes/molecular_computation/rdms.py

```diff
@@ -41,6 +48,96 @@ def matricize_2rdm(two_rdm, n_orbitals):
     return rho
 
 
+def compute_rdms(ferm_ham, exp_data, mapping, up_then_down):
+    """
+    Computes the 1- and 2-RDM and their spin-summed versions
+    using a Molecule object and frequency data either in the form of a
+    classical shadow or a dictionary of frequency histograms.
+
+    Args:
+        ferm_ham (FermionicOperator): Fermionic operator with n_spinorbitals and n_electrons defined
+        exp_data (ClassicalShadow or dict): classical shadow or a dictionary of expectation values of qubit terms
+        mapping: qubit mapping
+        up_then_down: spin ordering for the mapping
+
+    Returns:
+        complex array: 1-RDM
+        complex array: 2-RDM
+        complex array: spin-summed 1-RDM
+        complex array: spin-summed 2-RDM
+    """
+    onerdm = np.zeros((ferm_ham.n_spinorbitals,) * 2, dtype=complex)
+    twordm = np.zeros((ferm_ham.n_spinorbitals,) * 4, dtype=complex)
+    onerdm_spinsum = np.zeros((ferm_ham.n_spinorbitals//2,) * 2, dtype=complex)
+    twordm_spinsum = np.zeros((ferm_ham.n_spinorbitals//2,) * 4, dtype=complex)
+
+    # Check if the data dict is expectation values or raw frequency data in the {basis: hist} format
+    if isinstance(exp_data, dict) and isinstance(list(exp_data.values())[0], float):
+        exp_vals = {pauli_string_to_of(term): exp_val for term, exp_val in exp_data.items()}
+    elif isinstance(exp_data, dict) and isinstance(list(exp_data.values())[0], dict):
+        exp_vals = {}
+
+    n_qubits = get_qubit_number(mapping, ferm_ham.n_spinorbitals)
+
+    # Go over all terms in fermionic Hamiltonian
+    for term in ferm_ham.terms:
+        length = len(term)
+
+        if not term:
+            continue
+
+        # Fermionic term with a prefactor of 1.0.
+        fermionic_term = FermionOperator(term, 1.0)
+
+        qubit_term = fermion_to_qubit_mapping(fermion_operator=fermionic_term,
+                                            n_spinorbitals=ferm_ham.n_spinorbitals,
+                                            n_electrons=ferm_ham.n_electrons,
+                                            mapping=mapping,
+                                            up_then_down=up_then_down)
+        qubit_term.compress()
+
+        # Loop to go through all qubit terms.
+        eigenvalue = 0.
+
+        if isinstance(exp_data, RandomizedClassicalShadow):
+            for qterm, coeff in qubit_term.terms.items():
+                if coeff.real != 0:
+                    # Change depending on if it is randomized or not.
+                    eigenvalue += exp_data.get_term_observable(qterm, coeff)
+
+        if isinstance(exp_data, dict):
+            for qterm, coeff in qubit_term.terms.items():
+                if coeff.real != 0:
+
+                    try:
```

Comment:
I would change this to something more dense:
```python
exp_val = 1. if not qterm else exp_vals.get(term, some_function(...))
```
Where
```python
def some_function(...):
    ps = pauli_of_to_string(qterm, n_qubits)
    hist = aggregate_histograms(*[Histogram(exp_data[basis]) for basis in filter_bases(ps, 
    exp_data.keys())])
    return hist.get_expectation_value(qterm, 1.)
```

Reply:
Left as was, as per our discussion